src/librepgp/stream-parse.cpp

```diff
@@ -1438,7 +1442,9 @@ encrypted_try_key(pgp_source_encrypted_param_t *param,
         err = rsa_decrypt_pkcs1(
           &ctx.rng, decbuf.data(), &declen, &encmaterial.rsa, &keymaterial->rsa);
         if (err) {
-            RNP_LOG("RSA decryption failure");
+            if (!speculative) {
```

Comment:
What do you think on replacing all of these `if (!speculative)` with some define, like `RNP_LOG_IF(speculative, blablabla)`?
Also, as security context is passed everywhere, we may add property `rnp::SecurityContext.no_log()` instead of passing `speculative` parameter to every function call.

Reply:
> What do you think on replacing all of these if (!speculative) with some define, like RNP_LOG_IF(speculative, blablabla)?

Seems too big a response for such a local need.

Would it work if all these "decryption failed" messages emitted *by the library* could be considered debug-level messages and thus be suppressed by default? I would expect that if the error condition bubbles up to the application level, the application could show a message itself. The downside is that some details of "why" may be lost (to mitigate which we'd need to extend the enum of status codes), but indiscriminate logging of low-level intermediate unsuccessful outcomes is not a good UX either.

This would achieve us the following:
* no need to pass `bool speculative` down the stack
* don't need to remove or change much in the way the library currently emits stuff to stderr
* could possibly pave the way to the adoption of a conventional policy of "no printing by libraries"

Reply:
> Seems too big a response for such a local need. 

If rnp::SecurityContext.no_log() is an overkill, RNP_LOG_IF() would make code more readable, give less visual complexity. Also it would be easier to introduce changes in one place (for instance if we would like to add logging levels, whatever else).

> Would it work if all these "decryption failed" messages emitted by the library could be considered debug-level messages and thus be suppressed by default? 

Most of this logging is done to help to debug unexpected edge cases/configurations. I.e. we have problem which is already known then we'd report it to the caller, but in some cases caller need more information to understand what went wrong. 

However, we should not allow unlimited logging output as well - say, for instance, caller implemented something incorrectly and calls code which asks for the key in endless cycle. If log is redirected to the file then it will fill up the whole hard drive.

> could possibly pave the way to the adoption of a conventional policy of "no printing by libraries" 

User should be able to decide what to do with logs. We have issue #1158 related to this.

Reply:
Thanks for hinting at #1158. I totally agree with the argument from here that "Some appropriate logging context needs to be passed everywhere that logging is going to be done".

Regarding `SecurityContext.no_log()` and `RNP_LOG_IF()` i'd like to highlight that crypto backend functions such as `rsa_decrypt_pkcs1()` don't have neither `ffi` nor `SecurityContext` passed, yet OpenSSL-ish module logs into stderr via `RNP_LOG()`.

So we need to either pass that some context across the crypto backend boundary, or stop logs being emitted (by default; i think a special env var to enable is acceptable).

I am in favour of disabling, but I admit I have no insight into how bug reports get to us, how many, and how much back and forth with the reporters we can afford, and such.

Reply:
> Regarding `SecurityContext.no_log()` and `RNP_LOG_IF()` i'd like to highlight that crypto backend functions such as `rsa_decrypt_pkcs1()` don't have neither `ffi` nor `SecurityContext` passed, yet OpenSSL-ish module logs into stderr via `RNP_LOG()`. (Botan's module doesn't though, making it not problematic!)

Yeah, those redundant logging calls are added to be ready for the unexpected cases, as OpenSSL backend is comparable new and could be covered not that good as Botan. `SecurityContext` sooner or later will be passed to those functions, once we add possibility to mark certain asym algorithms and key sizes as weak, like this is done for hashes.

> 
> So we need to either pass that some context across the crypto backend boundary, or stop logs being emitted (by default; i think a special env var to enable is acceptable).

We already have `RNP_LOG_CONSOLE` which would disable logging at all. However, in this particular case, we need to disable logging from the certain function set, not from the whole library.

> I am in favour of disabling, but I admit I have no insight into how bug reports get to us, how many, and how much back and forth with the reporters we can afford, and such.

All reports (except security-sensitive ones) are created/added as Github issues here.

Reply:
> in this particular case, we need to disable logging from the certain function set, not from the whole library.

Could you please tell me which approaches to that would have your acceptance? Would make me happy if there are some easily fitting the scope of this task and PR.

After I wrote the above, I looked more closely into...

> We already have RNP_LOG_CONSOLE

And that was interesting. Changelog says

> Disabled logging by default in release builds and added support for environment variable `RNP_LOG_CONSOLE` to enable it back.

But that's not what I see. I've made a local release build, but the message from rsa_ossl.cpp is still being emitted. Is that a bug, or requirements changed since the changelog entry?

Reply:
@andrey-utkin Sorry, somehow forgot to reply here timely.
1. What do you think about the following solution to disable logging: https://github.com/rnpgp/rnp/pull/1799 ?
2. Hm, as from `logging.cpp`, for the release build (if NDEBUG is not defined) it should be set only if thereis non-zero RNP_LOG_CONSOLE enivornment variable.